### PR TITLE
feat(Header & Footer): Add header & footer to the dropdown list - iteration 2.

### PR DESCRIPTION
**Pull request by felixmosh**
This feature allows to add header & footer to the dropdowns for each of the default templates.

image

select2
selectize

### CODE REVIEW

**Comment by felixmosh** [comment]
Any updates?

**Comment by okolomiets** [comment]
The feature seems to become very expected one. Could anyone to review and merge it within next release? Thanks in advance.

**Comment by Jefiozie** [comment]
It is on the list. Only problem is I don't have a lot of time at the moment.
When I find the time I will do the review and then it will be there in next release.

File: src/common.css [hunk 1]
```diff
@@ -26,68 +25,84 @@
...
+  border-radius: 4px; /* FIXME hardcoded value :-/ */
+  border-bottom-left-radius: 0;
+  border-bottom-right-radius: 0;
 
-    border-top-width: 1px;  /* FIXME hardcoded value :-/ */
-    border-top-style: solid;
+  border-top-width: 1px; /* FIXME hardcoded value :-/ */
+  border-top-style: solid;
 
-    box-shadow: 0 -4px 8px rgba(0, 0, 0, 0.25);
```
Comment (Jefiozie) [inline src/common.css]: Can you change the spacing on this file? Looks like a lot of changes have been made but it are just spaces. 
thanks.

File: src/uiSelectFooterDirective.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  return {
+    templateUrl: function (tElement) {
+      // Needed so the uiSelect can detect the transcluded content
+      tElement.addClass('ui-select-footer');
+
+      // Gets theme attribute from parent (ui-select)
+      var theme = tElement.parent().attr('theme') || uiSelectConfig.theme;
+      return theme + '/footer.tpl.html';
+    },
+    restrict: 'EA',
```
Comment (Jefiozie) [inline src/uiSelectFooterDirective.js]: Should it match also the `attribute` tag? Or only the `element`?

File: src/uiSelectHeaderDirective.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  return {
+    templateUrl: function (tElement) {
+      // Needed so the uiSelect can detect the transcluded content
+      tElement.addClass('ui-select-header');
+
+      // Gets theme attribute from parent (ui-select)
+      var theme = tElement.parent().attr('theme') || uiSelectConfig.theme;
+      return theme + '/header.tpl.html';
+    },
+    restrict: 'EA',
```
Comment (Jefiozie) [inline src/uiSelectHeaderDirective.js]: Should it match also the `attribute` tag? Or only the `element`?

File: test/select.spec.js [hunk 4]
```diff
@@ -3378,4 +3378,52 @@ describe('ui-select tests', function () {
       expect(el.scope().$select.activeIndex).toBe(2);
     });
   });
+
+  describe('Header & Footer', function () {
+
+    var el;
+
+    function setupSelectComponent(theme, isMulti) {
```
Comment (Jefiozie) [inline test/select.spec.js]: We have 2 functions for building up the template `createUitSelect` and `createUiSelectMultiple`.  I believe this functions can be changed to support theses tests.

File: test/select.spec.js [hunk 5]
```diff
@@ -3378,4 +3378,52 @@ describe('ui-select tests', function () {
...
+          <ui-select-footer>{{' + selectedStr + '.name}}</ui-select-footer> \
+        </ui-select>'
+      );
+    }
+
+    ['selectize', 'bootstrap', 'select2'].forEach(function (theme) {
+      describe(theme + ' theme', function () {
+        it('should show the header', function () {
+          setupSelectComponent(theme);
+          expect($(el).find('.ui-select-header').text().trim()).toBe(scope.selection.selected.name);
```
Comment (Jefiozie) [inline test/select.spec.js]: Why are your trimming the text? Is there any reason for it?

Comment (felixmosh) [inline test/select.spec.js]: Cause it is a text in html element, and it can contain spaces...

File: src/uiSelectFooterDirective.js (hunk 2 above)
Comment (felixmosh) [inline src/uiSelectFooterDirective.js]: It can be both

File: src/uiSelectHeaderDirective.js (hunk 3 above)
Comment (felixmosh) [inline src/uiSelectHeaderDirective.js]: It can be both

File: src/common.css (hunk 1 above)
Comment (felixmosh) [inline src/common.css]: The spacing at the PR is the correct one, pay attention, there is inconstant spacing at the original file... i`ve just formatted it to be aligned.

File: test/select.spec.js (hunk 4 above)
Comment (felixmosh) [inline test/select.spec.js]: Yeah, but it doesn't have the `header` & `footer` tags...

**Comment by trosenfeld** [comment]
Just a comment:
As I was looking to use this feature I had made a local copy with these modifications.
When using the header and/or footer feature the options area in the dropdown no longer scrolls.
This is fairly annoying as the whole list of options is displayed which "elongates" the site when opening the dropdown. (We have lists of sometimes over 200 items)
You might also want to look into this when/if you get the time before releasing.

For now I'm just using the "master" ui-select and adding "disabled" items to the select items list using the "ui-disable-choice" attribute on "ui-select-choices" to prevent selection.

**Comment by felixmosh** [comment]
@trosenfeld, did u copied the styles as well?
U try to clone the PR, and run it locally.
I will be happy to look into your copy, to figure out what is wrong with it.

**Comment by felixmosh** [comment]
@Jefiozie any updates?

**Comment by GeneGenie** [comment]
So. when it will be available?

**Comment by felixmosh** [comment]
It depends on the author, he from some reason not approves it :/

**Comment by shyamal890** [comment]
@Jefiozie May be we should update this: https://github.com/angular-ui/ui-select/wiki/ui-select-header-and-footer

**Comment by bpaugam** [comment]
When will this be released ?

**Comment by besafx** [comment]
Can You Give Us A Code That We Can Add Into Select.JS to Enable This Feather

**Comment by pavel-blagodov** [comment]
Does anybody work on this repo? No commits since Mar 5

**Pull request merged**